storage/src/mmr/journaled.rs

```diff
@@ -0,0 +1,330 @@
+        let mut journal = Journal::<B, E, DIGEST_LENGTH>::init(context, journal_cfg).await?;
+        let mut journal_size = journal.size().await?;
+        if journal_size == 0 {
+            return Ok(Self {
+                mem_mmr: MemMmr::new(),
+                journal,
+                journal_size,
+            });
+        }
+
+        let zero_digest = H::Digest::try_from([0u8; DIGEST_LENGTH].as_ref()).unwrap();
+        let mut last_valid_size = journal_size;
+        while !PeakIterator::check_validity(last_valid_size) {
+            // Even this naive sequential backup must terminate in log2(n) iterations.
+            // A size-0 MMR is always valid so this loop must terminate before underflow.
+            last_valid_size -= 1;
+        }
+        let mut orphaned_leaf = zero_digest.clone();
+        if last_valid_size != journal_size {
+            warn!(
+                "encountered invalid MMR structure, recovering from last valid size of {}",
+                last_valid_size
+            );
+            // Check if there is an intact leaf following the last valid size, from which we can
+            // recover its missing parents.
+            let recovered_item = journal.read(last_valid_size).await;
+            if let Ok(item) = recovered_item {
+                orphaned_leaf = H::Digest::try_from(item.as_ref()).unwrap();
+            }
+            journal.rewind(last_valid_size).await?;
+            journal_size = last_valid_size
+        }
+
+        // TODO(https://github.com/commonwarexyz/monorepo/issues/478): extend Journal::replay() to
+        // accept a starting pos so this can be made much more efficient.
+        let mut vec = Vec::new();
+        let oldest_peak = PeakIterator::new(journal_size).next().unwrap().0;
+        for i in oldest_peak..journal_size {
+            let item = journal.read(i).await?;
+            vec.push(H::Digest::try_from(item.as_ref()).unwrap());
+        }
+        let mem_mmr = MemMmr::init(vec, oldest_peak);
+
+        let mut s = Self {
+            mem_mmr,
+            journal,
+            journal_size,
+        };
+
+        if orphaned_leaf != zero_digest {
+            // Recover the orphaned leaf and any missing parents.
+            let mut hasher = H::new();
+            let pos = s.add(&mut hasher, &orphaned_leaf);
+            assert!(pos == journal_size);
+            warn!("recovered orphaned leaf: {:?}", orphaned_leaf);
+            s.sync().await?;
+        }
+        Ok(s)
+    }
+
+    /// Add an element to the MMR and return its position in the MMR. Elements added to the MMR
+    /// aren't persisted to disk until `sync` is called.
+    pub fn add(&mut self, h: &mut H, element: &H::Digest) -> u64 {
+        self.mem_mmr.add(h, element)
+    }
+
+    pub fn root(&mut self, h: &mut H) -> H::Digest {
+        self.mem_mmr.root(h)
+    }
+
+    /// Sync any new elements to disk.
+    pub async fn sync(&mut self) -> Result<(), Error> {
```

Comment:
I wonder if it would be useful to allow the user to specify only to sync up to some pending index (rather than everything) 🤔 .

I think for now, its good to assume all